### DISCUSSION TITLE
From JATOS to Pavlovia with jsPsych?

### DISCUSSION THREAD

**Discussion by SchollJ** [jspsych/jsPsych#2458] [Q&A]
Dear jsPsych team,

I was wondering whether you could tell me whether it is feasible to do the following (it may sound like a strange request, I explain below why I need to do this):

I have an experiment that is compatible with JATOS without running jsPsych. In JATOS, to submit data, I give it my javascript structure in a single line. Is there anyway to (I'd of course initialise jsPsych etc.) pass a complete data structure to jsPsych at the end of an experiment in a similar way?

The reason I'm asking is that I have an experiment that runs on JATOS. Now we're applying for a new ethics and they are not happy with JATOS, but want us to use Pavlovia. However, I have not been able to find a simple way to submit data to Pavlovia, but I'm seeing on their website that there are jsPsych plugins to send the data to Pavlovia. So I was wondering whether I could maybe use jsPsych to get my data to Pavlovia, without having to recode the whole experiment to be in the format of jsPsych (or PsychoJS)?

I'd be very grateful for any pointers
Jacquie

**Comment by bjoluc** [comment]
Hi Jacquie,
you should be able to iterate over your existing list of result objects (assuming it's a flat list) and use [`jsPsych.data.get().push(result)`](https://www.jspsych.org/7.1/reference/jspsych-data/#push) for each result object before calling `jsPsych.run()`.

Cheers!

**EDIT:**
Sorry, I was assuming that each of your trials generates its own result data object. Since you only have one object (namely `data`) to store everything, and I realize you woud likely want to use jsPsych v6 (I think Pavlovia doesn't work with v7 out of the box), I would suggest that you use the [call-function](https://www.jspsych.org/6.3/plugins/jspsych-call-function/) plugin to do something like this after your experiment:

```js
var resultsTrial = {
  type: 'call-function',
  func: function() { return data; }
};

jsPsych.init({
  timeline: [resultsTrial, (pavlovia plugins go here) ]
})
```

**Reply by verenasarrazin** [reply]
Thank you very much for the proposed solution! Jacquie and I implemented it in our code and it is working. Instead of passing the whole data structure to the call function, we used a subset of the data and restructured it using _JSON.stringify_.

`var resultsTrial = {
    type: 'call-function',
    func: function(){
    var shortData = JSON.stringify(data.behaviorLongFormat)
    return shortData;}
};`

`jsPsych.init({
       timeline: [ resultsTrial, pavlovia_init, pavlovia_finish ]
  })`

And we also found out that we can add other parts of our data in the same way, by using a call function for each data structure and adding it to the timeline.

**Comment by SchollJ** [comment]
Thanks for the quick reply! Can I check what you mean by 'flat list' as I think mine is probably not flat, but I think I can potentially make it so. 

My data is like this: I have a structure called ‘data’ that throughout the experiment, I store data in. Below are just some examples what is stored in the structure. Then the data is usually ‘stringified’ and then send to JATOS. I’m wondering whether it might be possible to leave everything about the experiment the same and (after initialising PsychoJS of course), submit the data via PsychoJS?

```
# data.behaviour is a 2D matrix with participants choices on each trial and what was on the screen
data.behaviour.nSearches =....
data.behaviour.valueOnScreen=...

# in another part of the list, I'm storing experimental date and other information (e.g. questionnaire answers)
data.exptDate.endExpt = new Date();
data.exptDate.experimentDuration = data.exptDate.endExpt.getTime() - data.exptDate.startExpt.getTime();
data.exptDate.numberMcqAttempts = tSet.mcq.attemptCounter;
var resultJson = JSON.stringify(data);
jatos.submitResultData(resultJson)
```

**Reply by bjoluc** [reply]
Thanks for the clarification; I updated my answer!
